lib/ocpp/v201/smart_charging.cpp

```diff
@@ -402,4 +403,436 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
+
+    composite_schedule.evseId = evse_id;
+    composite_schedule.duration = SmartChargingHandler::determine_duration(start_time, end_time);
+    composite_schedule.scheduleStart = start_time;
+    composite_schedule.chargingRateUnit = charging_rate_unit;
+
+    return composite_schedule;
+}
+
+int32_t SmartChargingHandler::determine_duration(const ocpp::DateTime& start_time, const ocpp::DateTime& end_time) {
+    return duration_cast<seconds>(end_time.to_time_point() - start_time.to_time_point()).count();
+}
+
+bool SmartChargingHandler::within_time_window(const ocpp::DateTime& start_time, const ocpp::DateTime& end_time) {
+    return SmartChargingHandler::determine_duration(start_time, end_time) > 0;
+}
+
+PeriodDateTimePair SmartChargingHandler::find_period_at(const ocpp::DateTime& time, const ChargingProfile& profile,
+                                                        const int evse_id) {
+    auto period_start_time = this->get_profile_start_time(profile, time, evse_id);
+
+    EVLOG_info << "#" << profile.id << " find_period_at> " << period_start_time.value().to_rfc3339();
+
+    // TODO: Only processing first charging schedule
+    const auto schedule = profile.chargingSchedule.front();
+
+    if (period_start_time) {
+        const auto periods = schedule.chargingSchedulePeriod;
+        time_point<date::utc_clock> period_end_time;
+        for (size_t i = 0; i < periods.size(); i++) {
+            const auto period_end_time = get_period_end_time(i, period_start_time.value(), schedule);
+
+            EVLOG_info << "   find_period_at>        start_time> " << time.to_rfc3339();
+            EVLOG_info << "   find_period_at> period_start_time> " << period_start_time.value().to_rfc3339();
+            EVLOG_info << "   find_period_at>   period_end_time> " << period_end_time.to_rfc3339();
+
+            if (time >= period_start_time.value() && time < period_end_time) {
+                PeriodDateTimePair date_time_pair = {.period = periods.at(i),
+                                                     .end_time = ocpp::DateTime(period_end_time)};
+                log_period_date_time_pair(date_time_pair);
+                return date_time_pair;
+            }
+            period_start_time.emplace(ocpp::DateTime(period_end_time));
+        }
+    }
+
+    PeriodDateTimePair date_time_pair = {.period = std::nullopt, .end_time = MAX_DATE_TIME};
+    log_period_date_time_pair(date_time_pair);
+    return date_time_pair;
+}
+
+///
+/// \brief Calculates the composite schedule for the given \p valid_profiles and the given \p connector_id
+///
+/// Step 01-000 - Pass in profiles, start and end time, evse od add charging_rate_unit
+/// Step 01-001 - initialize_composite_schedule(start_time, end_time, evse_id, charging_rate_unit)
+/// Step 01-002 - Instantiate vector of ChargingSchedulePeriod periods
+CompositeSchedule SmartChargingHandler::calculate_composite_schedule(std::vector<ChargingProfile> valid_profiles,
```

Comment:
This is something I mentioned before: why take in `valid_profiles` instead of retrieving them from the `SmartChargingHandler` ourselves? That way we can know all of the profiles we're using to calculate a composite schedule are valid - allowing us to remove some checks that are further down.

Reply:
We can do this restructuring, but it will cause a significant delay to this merge. I would suggest making this a separate card, since the goal was to get this code in, and not devote a lot of time to restructuring. 

Personally, I would rather take the opposite approach and decouple the composite schedule logic altogether from the handler.